src/main/java/org/jenkinsci/plugins/workflow/cps/GlobalVariable.java

```diff
@@ -69,13 +70,26 @@
      * Gets or creates the singleton value of the variable.
      * If the object is stateful, and the state should not be managed externally (such as with a {@link RunAction2}),
      * then the implementation is responsible for saving it in the {@link CpsScript#getBinding}.
+     * <p>Implementors should specify a more concrete return type (making use of return type covariance).</p>
      * @param script the script we are running
      * @return a POJO or {@link GroovyObject}
      * @throws Exception if there was any problem creating it (will be thrown up to the script)
      * @see CpsScript#getProperty
      */
     public abstract @Nonnull Object getValue(@Nonnull CpsScript script) throws Exception;
 
+    /**
+     * Gets the logical type of this global variable - that is, the type returned by {@link #getValue}.
+     * <p>The default implementation returns the return type of {@link #getValue}, as implemented in the subclass.</p>
+     */
+    public Type getType() {
```

Comment:
This is normally done by making the class generic and adding a method returning `Class<T>`.

Reply:
The idea, iirc after 3 years, was to provide proper type information for tooling; eg, an IDE plugin backed by a real Jenkins instance or Jenkins Test Harness. I chose `Type` over `Class<T>` because `Type` can fully represent types including generics etc.

That said, I've moved on from this, and it's unlikely that I'll work on it in the foreseeable future. Please feel free to close this and the other PR. Sorry for the mess/confusion.

Reply:
> an IDE plugin backed by a real Jenkins instance

Like GDSL as in https://github.com/jenkinsci/workflow-step-api-plugin/pull/40#pullrequestreview-189529757?